### DISCUSSION TITLE
Orbital Parallax?

### DISCUSSION THREAD

**Discussion by prideout** [cosinekitty/astronomy#239] [General]
Don, I'm using your awesome library for its `Astronomy_Horizon` function and I noticed that it accounts for topocentric parallax.  This helps when dealing with nearby objects like the moon.  However I _think_ it does not account for parallax due to the Earth's orbit around the sun.  This might make a small difference if I'm using the function for nearby stars, or other extra-solar objects.

If I'm wrong and it _does_ account for orbital parallax, can you point me to the codelines for it?  I'm curious about the math.

Thanks again for a great library!  🚀

**Comment by cosinekitty** [comment]
Hi again @prideout! Good to hear from you again. Currently Astronomy Engine focuses on solar system astronomy. I don't currently have any functions for directly calculating apparent positions of objects far from the solar system. However, if you know the equatorial coordinates (RA, DEC) and distance (in AU) of a star in heliocentric terms, I can show you how to use Astronomy Engine to correct its position for an observer on the Earth. Geocentric will be perfectly accurate; there is no conceivable way to measure any difference between topocentric and geocentric parallax for objects that far away. Consider that 1 light year = 63241 AU, and the radius of the earth is 1/23450 of an AU. That makes a really skinny triangle, LOL!

Also, correcting for [aberration](https://en.wikipedia.org/wiki/Aberration_(astronomy)) is generally orders of magnitude more important than parallax (22 arcseconds max, versus less than 1 arcsecond even for the closest extrasolar stars), which Astronomy Engine also supports.

I think in practice you will find correcting for [proper motion](https://en.wikipedia.org/wiki/Proper_motion) of a star over time is also more important than parallax for obtaining an accurate position. There are several star catalogs out there that include positions and proper motion data for thousands of stars.

Let me know if you are interested in more specifics and I can help point you in the right direction.

**Reply by prideout** [reply]
Thanks Don!  Everything you said makes sense to me.

As you say, the library focuses on Solar System bodies, so the `Horizon` and `Equator` functions take Solar System bodies, not arbitrary coordinates.

However it shouldn't be too hard to use the library facilities to, say, convert arbitrary ICRF coordinates into true-of-date equatorial coordinates for a particular observer at a particular time.

Here is my [wrong] attempt to do so:

```python
def ICRF_to_TOD(ra, dec, distance_from_origin, time, observer):
    """
    Converts ICRF coordinates into true-of-date equatorial coordinates.
    This produces what Horizons calls "apparent" coordinates.
    """
    spherical_icrf = astro.Spherical(lat=dec, lon=ra, dist=distance_from_origin)
    cartesian_icrf = astro.VectorFromSphere(spherical_icrf, time)
    earth_ssb = astro.BaryState(astro.Body.Earth, time)
    cartesian_geocentric = cartesian_icrf - earth_ssb
    return EquatorCoords(cartesian_geocentric, time, observer, True)

def EquatorCoords(gc_body, time, observer, ofdate):
    """
    Similar to Equator(), but takes geocentric coordinates rather than a body.
    """
    gc_observer = astro.ObserverVector(time, observer, True)
    j2000 = [
        gc_body.x - gc_observer.x,
        gc_body.y - gc_observer.y,
        gc_body.z - gc_observer.z
    ]
    if not ofdate:
        return astro._vector2radec(j2000, time)
    temp = astro._precession(j2000, time, astro._PrecessDir.From2000)
    datevect = astro._nutation(temp, time, astro._PrecessDir.From2000)
    return astro._vector2radec(datevect, time)
```

However I compared the results of the above code with what Horizons produces and it's way off, so I'm doing something very wrong.

For my test I'm using the coordinates of the Ceres Asteroid on 2022 Jan 1, which has ICRF coords ra=56.57196,dec=17.72584.  According to Horizons, this should result in ra=56.88681 dec=17.79389 for the true-of-date equatorial coordinates from my location in San Jose, CA.  Instead the above code produces ra=2.703 dec=12.921 😆

**Reply by cosinekitty** [reply]
Hi Philip! I'm kinda slammed today, but I will take a look at this over the weekend and get back to you. I'm sure I can help.

**Reply by cosinekitty** [reply]
Your approach looks right overall. One thing that could be a little tricky is whether right ascension is expressed in sidereal hours (0..24) or degrees (0..360). JPL Horizons looks like it's using degrees, so passing `lon=ra` to the `Spherical` class constructor is correct. That constructor does expect degrees.

However, `_vector2radec` returns right ascension in sidereal hours. Multiplying 2.703*15 = 40.545 degrees.

I tried to reproduce what you are doing, and I don't get the same numbers from JPL Horizons, which means we must be using different settings. I tried calculating the barycentric position of Ceres at midnight UTC on 2022 Jan 1.  I get:  RA = 71.24365 degrees, DEC = 20.82891 degrees, distance = 2.71889860021381 AU. My settings look like this:

Ephemeris Type = Observer Table
Target Body = 1 Ceres (A801 AA)
Observer Location: Solar System Barycenter (SSB) [500@0]
Time Specification: Start=2022-01-01 00:00 UT
Table Settings: I checked boxes 1 (Astrometric RA & DEC) and 20 (Observer range & range-rate).
Reference frame = ICRF
Angle format = decimal degrees
Refraction model = no refraction (airless)
Range units = astronomical units (au)

Then I switch the settings for the coordinates of San Jose:
Observer Location: San Jose, CA (121°53′24.0″W, 37°20′16.1″N)

This results in observer-centered ICRF coordinates:
RA=56.57196 degrees, DEC=17.72584 degrees, distance = 1.90625754381826 AU.

But those are still coordinates using the J2000 equator, not the 2022 equator. I can't see an option in JPL Horizons to convert from ICRF to of-date coordinates. How did you do that?

**Reply by prideout** [reply]
Horizons can generate something called "Apparent Coordinates" which I was interpreting as "of-date" but their User Manual isn't terribly clear.  It defines Apparent Coordinates as:

> Apparent coordinates can be with respect to an inertial frame such as ICRF or FK4/B1950, such as for space-based observers (spacecraft) or, for observers on a rotating surface, with respect to some “of-date” coordinate system, involving precession-nutation to the Earth’s true-equator and equinox-of-date (or that of a non-Earth body associated with the observer location).


Attached are screenshots of the settings I used.  Thanks for working with me on this :)

<img width="762" alt="page1" src="https://user-images.githubusercontent.com/1288904/189502864-bdaf4819-3718-4624-b721-8f6bdbb30ea9.png">

<img width="753" alt="page2" src="https://user-images.githubusercontent.com/1288904/189502866-d9750cc9-c992-42a1-96f2-797f6335b705.png">

**Reply by cosinekitty** [reply]
I'm pretty sure the only difference between checkbox 1 (astrometric) and 2 (apparent) is whether atmospheric refraction and/or aberration corrections are applied. So let's back up a moment to talk about what you're really trying to do. Is there a reason you want to find coordinates relative to the Earth's equator-of-date? I'm not saying you shouldn't want that, I'm just trying to clarify. The main reason someone would need of-date coordinates is to accurately calculate horizontal coordinates for an observer. Then you really do need to know where a celestial object is in relation to the Earth's equator and axis at that moment in time. But other times, it's easier to use a fixed orientation system for locating e.g. a planet relative to the stars in nearby constellations.

**Reply by prideout** [reply]
Yes, I'd like to find the equator-of-date coordinates so that I can either (1) obtain accurate horizontal coordinates for an observer, or (2) obtain highly accurate equatorial coordinates for an observer who has a telescope with an equatorial mount.

It sounds like my code might actually be correct (with the added HMS => Deg conversion that you mentioned), but that Horizons isn't doing what I expected.  Interesting!  Maybe I should try comparing against a system other than Horizons, since I'm not piloting a spacecraft... :)

**Reply by cosinekitty** [reply]
OK, that makes sense. Just as an example, I have spotted Vesta in my 8-inch reflector. I found it by dialing in geocentric coordinates into my old-school mechanical equatorial mount. I did not worry at all about correcting for topocentric parallax, because the error isn't enough to cause you to miss spotting the object. In fact, you can still nail the object right in the center of your scope if you do it right.

For an object 2 AU away (like Ceres), we can calculate the maximum angular error by thinking about how large the Earth appears to an observer on Ceres. In other words, the change in the sky position of Ceres for observers on opposite sides of the Earth is exactly the same as the angular diameter of the Earth as seen by an observer on Ceres.

Divide the diameter of the Earth in AU = 8.527e-05 by the distance 2 AU to get 4.264e-05 radians. This converts to 8.8 arcseconds, which is tiny compared to the field of view through even a high-power objective lens. If you have seen Mars through a telescope, the size it appears gives you an idea of how much a geocentric/topocentric correction will matter for an asteroid.

So if all you want is equatorial coordinates, just use geocentric coordinates however you can get them. The Moon is the only solar system body close enough for a topocentric correction to make a significant difference for backyard astronomers.

For horizontal coordinates, here is what I would encourage you to try:

1. Use VectorFromSphere, like you already did, to convert geocentric J2000 equatorial coordinates into a vector.
2. Call [Rotation_EQJ_HOR](https://github.com/cosinekitty/astronomy/tree/master/source/python#Rotation_EQD_HOR) to calculate the rotation matrix that converts equatorial vectors into horizontal vectors for a given time and observer.
3. Pass the rotation matrix and the equatorial vector to [RotateVector](https://github.com/cosinekitty/astronomy/tree/master/source/python#RotateVector) to calculate a horizontal vector.
4. Pass the horizontal vector to [VectorFromHorizon](https://github.com/cosinekitty/astronomy/tree/master/source/python#VectorFromHorizon) to convert the vector into altitude and azimuth angles.

Sounds like a fun project! And I hope this helps.

Also, if you are interested, Astronomy Engine does provide a gravity simulator that can extrapolate position and velocity vectors for small bodies like asteroids and comets, given an initial seed state within a year or two of the time you are interested in. See the [GravitySimulator](https://github.com/cosinekitty/astronomy/tree/master/source/python#GravitySimulator) class, and take a look at the unit test `generate/test.py` for the function `GravSim`. The place where I test calculating Ceres is

```python
if 0 != GravSimFile("geostate/Ceres.txt",     astronomy.Body.Earth, nsteps, 6.5689, 6.4797): return 1
```

Let me know if you would like some help working with that, and I can give you more information about it.

**Reply by prideout** [reply]
Hi Don, your suggestion works nicely, here's what I wrote:

```python
def ICRF_to_Horizontal(ra, dec, distance, time, observer):
    barycentric_spherical = astro.Spherical(lat=dec, lon=ra, dist=distance)
    barycentric_vector = astro.VectorFromSphere(barycentric_spherical, time)

    # The following line is just an approximation.
    geocentric_vector = barycentric_vector

    matrix = astro.Rotation_EQJ_HOR(time, observer)
    horizontal_vector = astro.RotateVector(matrix, geocentric_vector)
    horizontal_spherical = astro.HorizonFromVector(horizontal_vector, astro.Refraction.Airless)
    az = horizontal_spherical.lon
    alt = horizontal_spherical.lat
    return alt, az

```

This produces `alt=19.995 az=82.438` which is pretty close to Horizons `(alt=19.991 az=82.434)`.

I wonder if these differences mostly arise from the fact that ICRF defines origin=SSB whereas the code above assumes that origin=geocenter.  It doesn't matter for backyard astronomy, but enjoy learning about all the effects that apply to "apparent" coordinates.

To test my theory, I tried replacing the `geocentric_vector = barycentric_vector` approximation with:

```python
earth_barycentric_position = astro.BaryState(astro.Body.Earth, time)
geocentric_vector = barycentric_vector - earth_barycentric_position
```

Alas, this produces way-off results. I think that maybe Horizon's reported "Astrometric ICRF" coordinate is actually already geocentric... if so that's a bit confusing since ICRF is formally defined with its origin is at the barycenter of the Solar System.

Anyway -- thanks a lot, you helped clarify a lot of things!

**Reply by cosinekitty** [reply]
That's great! I'm glad this was helpful. And, yes, the reference frames can have an arbitrary origin; they are meant to define direction vectors in space but don't require a particular origin. So ICRF can be heliocentric, barycentric, geocentric, topocentric, or whatever, depending on context. Same thing for other reference frames: ecliptic, equator-of-date, galactic.

In this case, the origin is geocentric for sure. Treating it as barycentric causes the huge errors because your target object Ceres is not very far away (~ 2AU) compared to the distance between the Earth and the Solar System Barycenter (~ 1AU). So Ceres appears in a very different position against distant background stars when viewed from the SSB than when viewed from Earth.

When you use Horizon's intended geocentric origin, the small discrepancy is mostly due to Astronomy Engine using simplified models of precession and nutation of the Earth's axis over time. If I used the highest precision possible, it would greatly bloat the size of the code and make it run much slower, for a very tiny improvement of accuracy. Horizons also has some gnarly relativistic corrections, while Astronomy Engine is strictly Newtonian. Their accuracy goal is sub-milli-arcsecond; mine are arcminute, so the resulting code can be much smaller and faster than theirs.

**Reply by cosinekitty** [reply]
Oh, correction... your screenshot shows you chose a topocentric location (San Jose, CA) for the observer. If I understand what you did to get your numbers, that's your origin, which would account for a few arcseconds of discrepancy also. That could easily affect the third digit after the decimal point.
